File: lobster-input-android/app/src/main/java/com/lobster/input/keyboard/typing/coordinator/TypingSessionHost.kt
package com.lobster.input.keyboard.typing.coordinator

import android.content.Context
import com.lobster.input.core.locale.MobileLanguage
import com.lobster.input.core.locale.MobileStrings
import com.lobster.input.keyboard.KeyboardRealtimeMic
import com.lobster.input.keyboard.pinyin.PinyinLoadState
import com.lobster.input.keyboard.typing.InputLang
import com.lobster.input.keyboard.typing.KbStr
import com.lobster.input.keyboard.typing.TypingKeyboardHost
import com.lobster.input.keyboard.typing.composing.ComposingTextBridge
import com.lobster.input.keyboard.typing.composing.TypingUndoManager
import com.lobster.input.keyboard.typing.editor.EditorActionModel
import com.lobster.input.keyboard.typing.editor.EditorInfoResolver

/**
 * [TypingKeyboardHost] 实现:上屏、组合预览、Enter 动作、撤销、本地化。
 */
class TypingSessionHost(
    private val context: Context,
    private val ime: TypingImeContext,
    private val preferences: TypingPreferences,
    private val composingBridge: ComposingTextBridge,
    private val undoManager: TypingUndoManager,
    private var editorAction: EditorActionModel = EditorActionModel.DEFAULT,
    private var statusCallback: ((String?) -> Unit)? = null,
    private var inputBlockedCallback: ((Boolean) -> Unit)? = null
) : TypingKeyboardHost {

    fun refreshEditorAction() {
        editorAction = EditorInfoResolver.resolve(context, ime.editorInfo())
    }

    fun setStatusCallback(callback: (String?) -> Unit) {
        statusCallback = callback
    }

    fun setInputBlockedCallback(callback: (Boolean) -> Unit) {
        inputBlockedCallback = callback
    }

    override fun onCommit(text: String) {
        composingBridge.clear()
        ime.inputConnection()?.commitText(text, 1)
        undoManager.record(text)
    }

    // 韩语组字预览:组合中音节直接写目标框(setComposingText,业界标准的光标下划线预览)
    override fun onComposingText(text: String) {
        ime.inputConnection()?.setComposingText(text, 1)
    }

    override fun onFinishComposing() {
        ime.inputConnection()?.finishComposingText()
    }

    override fun onCommitPair(opening: String, closing: String) {
        composingBridge.clear()
        val ic = ime.inputConnection()
        if (ic == null) { undoManager.record(opening); return }
        ic.beginBatchEdit()
        ic.commitText(opening + closing, 1)
        val pos = ic.getTextBeforeCursor(Int.MAX_VALUE, 0)?.length ?: (opening.length + closing.length)
        ic.setSelection(pos - closing.length, pos - closing.length)
        ic.endBatchEdit()
        // 撤销时按整体回退;记录两字符长度
        undoManager.record(opening + closing)
    }

    override fun hasTextAfterCursor(): Boolean =
        ime.inputConnection()?.getTextAfterCursor(1, 0)?.isNotEmpty() == true

    override fun textBeforeCursor(maxChars: Int): String =
        ime.inputConnection()?.getTextBeforeCursor(maxChars, 0)?.toString() ?: ""

    override fun onMoveCursor(dx: Int) {
        val ic = ime.inputConnection() ?: return
        if (dx == 0) return
        val pos = ic.getTextBeforeCursor(Int.MAX_VALUE, 0)?.length ?: return
        val after = ic.getTextAfterCursor(Int.MAX_VALUE, 0)?.length ?: 0
        val target = (pos + dx).coerceIn(0, pos + after)
        ic.setSelection(target, target)
    }

    override fun onDeleteBeforeCursor() {
        val ic = ime.inputConnection() ?: return
        val selected = ic.getSelectedText(0)
        if (!selected.isNullOrEmpty()) ic.commitText("", 1) else ic.deleteSurroundingText(1, 0)
    }

    override fun onEnter() {
        composingBridge.clear()
        val ic = ime.inputConnection() ?: return
        when (editorAction.action) {
            EditorActionModel.EditorInfoAction.SEARCH -> {
                ic.performEditorAction(android.view.inputmethod.EditorInfo.IME_ACTION_SEARCH)
            }
            EditorActionModel.EditorInfoAction.SEND -> {
                ic.performEditorAction(android.view.inputmethod.EditorInfo.IME_ACTION_SEND)
            }
            EditorActionModel.EditorInfoAction.GO -> {
                ic.performEditorAction(android.view.inputmethod.EditorInfo.IME_ACTION_GO)
            }
            EditorActionModel.EditorInfoAction.DONE -> {
                ic.performEditorAction(android.view.inputmethod.EditorInfo.IME_ACTION_DONE)
            }
            EditorActionModel.EditorInfoAction.NEXT -> {
                ic.performEditorAction(android.view.inputmethod.EditorInfo.IME_ACTION_NEXT)
            }
            EditorActionModel.EditorInfoAction.PREVIOUS -> {
                ic.performEditorAction(android.view.inputmethod.EditorInfo.IME_ACTION_PREVIOUS)
            }
            EditorActionModel.EditorInfoAction.NEWLINE -> {
                if (!ic.commitText("\n", 1)) {
                    val t = System.currentTimeMillis()
                    ic.sendKeyEvent(android.view.KeyEvent(t, t, android.view.KeyEvent.ACTION_DOWN, android.view.KeyEvent.KEYCODE_ENTER, 0))
                    ic.sendKeyEvent(android.view.KeyEvent(t, t, android.view.KeyEvent.ACTION_UP, android.view.KeyEvent.KEYCODE_ENTER, 0))
                }
            }
        }
    }

    override fun onReturnToVoice() = ime.requestExitTypingMode()

    override fun onKeyboardMicStart() { /* 由 KeyboardMicCoordinator 处理 */ }

    override fun onKeyboardMicStop() { /* 由 KeyboardMicCoordinator 处理 */ }

    override fun isChineseModeDefault(): Boolean = preferences.chineseMode

    override fun persistChineseMode(chinese: Boolean) {
        preferences.chineseMode = chinese
    }

    override fun inputLangDefault(): InputLang = InputLang.fromCode(preferences.inputLang)

    override fun persistInputLang(lang: InputLang) {
        preferences.inputLang = lang.code
        // 中英偏好联动(向后兼容旧逻辑读取点);俄/韩不动 chineseMode
        if (lang == InputLang.ZH || lang == InputLang.EN) preferences.chineseMode = lang == InputLang.ZH
    }

    /** App 界面语言 → 语言组(中英组含粤语/繁体)与组默认输入语种。 */
    private fun appLangGroup(): Pair<String, InputLang> = when (MobileStrings.currentLanguage(context)) {
        MobileLanguage.RU -> "ru" to InputLang.RU
        MobileLanguage.KO -> "ko" to InputLang.KO
        MobileLanguage.EN -> "zhen" to InputLang.EN
        else -> "zhen" to InputLang.ZH // zh / zh-Hant / yue 均属中文
    }

    override fun groupDefaultLang(): InputLang = appLangGroup().second

    override fun consumeLangGroupChange(): Boolean {
        val (group, def) = appLangGroup()
        val last = preferences.langGroup
        preferences.langGroup = group
        // 首次(无记录):中英组不算切换;俄/韩界面首启也要落到对应语种键盘
        if (last == null && group != "zhen") {
            preferences.inputLang = def.code
            return true
        }
        // 组内互切(中英互换界面语言)不清暂存
        if (last == null || last == group) return false
        // 跨组切换:清空暂存、恢复初始化——输入语种重置为新组默认
        // (zh/繁/粤界面→中文,en界面→英文,ru→俄语,ko→韩语),中英偏好同步对齐
        preferences.inputLang = def.code
        if (def == InputLang.ZH || def == InputLang.EN) preferences.chineseMode = def == InputLang.ZH
        return true
    }

    override fun isNineGridDefault(): Boolean = preferences.nineGrid

    override fun persistNineGrid(nine: Boolean) {
        preferences.nineGrid = nine
    }

    override fun onKeyFeedback() = ime.performKeyHaptic()

    override fun localize(key: KbStr): String = when (key) {
        KbStr.RETURN_VOICE -> MobileStrings.returnVoice(context)
        KbStr.SPACE -> MobileStrings.spaceKey(context)
        KbStr.CH -> MobileStrings.imeChineseLabel(context)
        KbStr.EN -> MobileStrings.imeEnglishLabel(context)
        KbStr.NINEGRID -> MobileStrings.nineGridLabel(context)
        KbStr.QWERTY -> MobileStrings.qwertyLabel(context)
        KbStr.SYMBOL -> MobileStrings.symbolKeyLabel(context)
        KbStr.SWITCH_QWERTY -> MobileStrings.switchQwertyLabel(context)
        KbStr.PINYIN -> MobileStrings.pinyinKeyLabel(context)
        KbStr.ABC -> MobileStrings.abcKeyLabel(context)
        KbStr.SYLLABLE -> MobileStrings.syllableSepKey(context)
        KbStr.UNDO -> MobileStrings.undo(context)
    }

    override fun enterKeyLabel(): String = editorAction.label

    override fun isPasswordField(): Boolean = editorAction.isPassword

    override fun isNumericField(): Boolean = editorAction.isNumeric

    override fun isEnglishField(): Boolean = editorAction.isEnglish

    override fun isMicAvailable(): Boolean = ime.isLoggedIn()

    fun canRememberTypingMode(): Boolean = preferences.rememberTypingMode

    fun lastModeKeyboard(): Boolean = preferences.lastModeKeyboard
    fun setLastModeKeyboard(value: Boolean) { preferences.lastModeKeyboard = value }

    fun fuzzySettings() = preferences.fuzzySettings()

    override fun recordRecentEmoji(emoji: String) = preferences.recordRecentEmoji(emoji)

    override fun recentEmojis(): List<String> = preferences.recentEmojis()

    override fun recordRecentSymbol(symbol: String) = preferences.recordRecentSymbol(symbol)

    override fun recentSymbols(): List<String> = preferences.recentSymbols()

    private val clipboardStore = ClipboardStore(context)

    // 密码框不读剪贴板(隐私)
    override fun clipboardHistory(): List<String> =
        if (isPasswordField()) emptyList() else clipboardStore.snapshotAndGet()

    override fun quickPhrases(): List<String> =
        if (isPasswordField()) emptyList() else clipboardStore.quickPhrases()

    override fun removeClipboardItem(item: String) = clipboardStore.removeHistory(item)

    override fun clearClipboard() = clipboardStore.clearHistory()

    fun updateComposingPreview(text: String) = composingBridge.update(text)

    fun clearComposingPreview() = composingBridge.clear()

    fun showStatus(message: String?) = statusCallback?.invoke(message)

    fun setInputBlocked(blocked: Boolean) = inputBlockedCallback?.invoke(blocked)

    override fun canUndo(): Boolean = undoManager.canUndo()

    override fun undoLastCommit() {
        val last = undoManager.pop() ?: return
        val ic = ime.inputConnection() ?: return
        ic.deleteSurroundingText(last.length, 0)
        // 不再弹"已撤销"状态提示:它会让候选栏多出一行、顶高输入面板,体验差
    }

    fun notifyEngineState(state: PinyinLoadState, detail: String?) {
        val msg = when (state) {
            PinyinLoadState.LOADING -> MobileStrings.pinyinDictLoading(context)
            PinyinLoadState.FAILED -> detail ?: MobileStrings.pinyinDictFailed(context)
            PinyinLoadState.READY -> null
            PinyinLoadState.IDLE -> null
        }
        showStatus(msg)
        setInputBlocked(state == PinyinLoadState.LOADING || state == PinyinLoadState.FAILED)
    }
}

File: lobster-input-ios/LobsterKeyboard/Typing/Coordinator/TypingSessionHost.swift
import UIKit

/// [TypingKeyboardHost] 实现:上屏、组合预览、Enter 动作、撤销、本地化。
final class TypingSessionHost: TypingKeyboardHost {

    /// weak 而非 unowned:连删 Timer 等异步回调可能在键盘 VC 销毁后的间隙触达这里,
    /// unowned 会直接 abort(真机已复现闪退);weak + 优雅降级为 no-op 才安全。
    private weak var ime: TypingImeContext?
    private let preferences: TypingPreferences
    private let composingBridge: ComposingTextBridge
    private let undoManager: TypingUndoManager
    private var editorAction: EditorActionModel = .default
    private var statusCallback: ((String?) -> Void)?
    private var inputBlockedCallback: ((Bool) -> Void)?

    init(
        ime: TypingImeContext,
        preferences: TypingPreferences,
        composingBridge: ComposingTextBridge,
        undoManager: TypingUndoManager
    ) {
        self.ime = ime
        self.preferences = preferences
        self.composingBridge = composingBridge
        self.undoManager = undoManager
        composingBridge.bind(
            proxyProvider: { [weak ime] in ime?.textDocumentProxy },
            micActiveProvider: { [weak ime] in ime?.isMicRecording() ?? false }
        )
    }

    func refreshEditorAction() {
        guard let ime else { return }
        editorAction = EditorActionModel.resolve(from: ime.textDocumentProxy)
    }

    func setStatusCallback(_ callback: @escaping (String?) -> Void) {
        statusCallback = callback
    }

    func setInputBlockedCallback(_ callback: @escaping (Bool) -> Void) {
        inputBlockedCallback = callback
    }

    func onCommit(_ text: String) {
        guard let ime else { return }
        composingBridge.clear()
        ime.textDocumentProxy.insertText(text)
        undoManager.record(text)
    }

    func onCommitPair(_ opening: String, _ closing: String) {
        guard let ime else { return }
        composingBridge.clear()
        let proxy = ime.textDocumentProxy
        proxy.insertText(opening + closing)
        proxy.adjustTextPosition(byCharacterOffset: -closing.count)
        undoManager.record(opening + closing)
    }

    /// 光标后是否还有文字(1 键符号候选的成套插入判定)。
    func hasTextAfterCursor() -> Bool {
        guard let ime else { return false }
        return !(ime.textDocumentProxy.documentContextAfterInput ?? "").isEmpty
    }

    /// 光标前 ≤maxChars 字符(标点联想 v2 句内上下文;documentContextBeforeInput 为 nil 时返回空)。
    func textBeforeCursor(_ maxChars: Int) -> String {
        guard let ime else { return "" }
        let ctx = ime.textDocumentProxy.documentContextBeforeInput ?? ""
        return String(ctx.suffix(maxChars))
    }

    func onDeleteBeforeCursor() {
        guard let ime else { return }
        let proxy = ime.textDocumentProxy
        if let sel = proxy.selectedText, !sel.isEmpty {
            proxy.insertText("")
        } else {
            proxy.deleteBackward()
        }
    }

    func onEnter() {
        guard let ime else { return }
        composingBridge.clear()
        let proxy = ime.textDocumentProxy
        switch editorAction.action {
        case .search, .send, .go, .done, .next, .previous:
            proxy.insertText("\n")
        case .newline:
            proxy.insertText("\n")
        }
    }

    func onReturnToVoice() { ime?.requestExitTypingMode() }

    func onKeyboardMicStart() { /* 由 TypingMicCoordinator 处理 */ }

    func onKeyboardMicStop() { /* 由 TypingMicCoordinator 处理 */ }

    func isChineseModeDefault() -> Bool { preferences.chineseMode }

    func persistChineseMode(_ chinese: Bool) { preferences.chineseMode = chinese }

    func inputLangDefault() -> InputLang { InputLang(rawValue: preferences.inputLang) ?? .zh }

    func persistInputLang(_ lang: InputLang) {
        preferences.inputLang = lang.rawValue
        // 中英偏好联动(向后兼容旧逻辑读取点);俄/韩不动 chineseMode
        if lang == .zh || lang == .en { preferences.chineseMode = lang == .zh }
    }

    /// App 界面语言 → 语言组(中英组含粤语/繁体)与组默认输入语种。
    private func appLangGroup() -> (String, InputLang) {
        switch MobileStrings.language {
        case .ru: return ("ru", .ru)
        case .ko: return ("ko", .ko)
        case .en: return ("zhen", .en)
        default: return ("zhen", .zh) // zh / zh-Hant / yue 均属中文
        }
    }

    func groupDefaultLang() -> InputLang { appLangGroup().1 }

    func consumeLangGroupChange() -> Bool {
        let (group, def) = appLangGroup()
        let last = preferences.langGroup
        preferences.langGroup = group
        // 首次(无记录):中英组不算切换;俄/韩界面首启也要落到对应语种键盘
        if last == nil && group != "zhen" {
            preferences.inputLang = def.rawValue
            return true
        }
        // 组内互切(中英互换界面语言)不清暂存
        if last == nil || last == group { return false }
        // 跨组切换:清空暂存、恢复初始化——输入语种重置为新组默认,中英偏好同步对齐
        preferences.inputLang = def.rawValue
        if def == .zh || def == .en { preferences.chineseMode = def == .zh }
        return true
    }

    // 韩语组字预览:组合中音节写目标框 marked 区(setMarkedText,业界标准预览)
    func onComposingText(_ text: String) {
        ime?.textDocumentProxy.setMarkedText(text, selectedRange: NSRange(location: text.count, length: 0))
    }

    func onFinishComposing() {
        ime?.textDocumentProxy.unmarkText()
    }

    func recordRecentSymbol(_ symbol: String) { preferences.recordRecentSymbol(symbol) }

    func recentSymbols() -> [String] { preferences.recentSymbols() }

    func isNineGridDefault() -> Bool { preferences.nineGrid }

    func persistNineGrid(_ nine: Bool) { preferences.nineGrid = nine }

    func onKeyFeedback() { ime?.performKeyHaptic() }

    func localize(_ key: KbStrKey) -> String {
        switch key {
        case .returnVoice: return MobileStrings.returnVoice()
        case .space: return MobileStrings.spaceKey()
        case .ch: return MobileStrings.imeChineseLabel()
        case .en: return MobileStrings.imeEnglishLabel()
        case .ninegrid: return MobileStrings.nineGridLabel()
        case .qwerty: return MobileStrings.qwertyLabel()
        case .symbol: return MobileStrings.symbolKeyLabel()
        case .switchQwerty: return MobileStrings.switchQwertyLabel()
        case .pinyin: return MobileStrings.pinyinKeyLabel()
        case .abc: return MobileStrings.abcKeyLabel()
        case .syllable: return MobileStrings.syllableSepKey()
        case .undo: return MobileStrings.undo()
        }
    }

    func enterKeyLabel() -> String { editorAction.label }

    func isPasswordField() -> Bool { editorAction.isPassword }

    func isNumericField() -> Bool { editorAction.isNumeric }

    func isEnglishField() -> Bool { editorAction.isEnglish }

    func updateComposingPreview(_ text: String) { composingBridge.update(composingDisplay: text) }

    func clearComposingPreview() { composingBridge.clear() }

    func showStatus(_ message: String?) { statusCallback?(message) }

    func setInputBlocked(_ blocked: Bool) { inputBlockedCallback?(blocked) }

    func canUndo() -> Bool { undoManager.canUndo() }

    func undoLastCommit() {
        guard let ime, let last = undoManager.pop() else { return }
        for _ in last { ime.textDocumentProxy.deleteBackward() }
        // 不再弹"已撤销"状态提示:会让候选栏多一行、顶高输入面板
    }

    func notifyEngineState(_ state: PinyinLoadState, detail: String?) {
        let msg: String?
        switch state {
        case .loading: msg = MobileStrings.pinyinDictLoading()
        case .failed: msg = detail ?? MobileStrings.pinyinDictFailed()
        case .ready, .idle: msg = nil
        }
        showStatus(msg)
        setInputBlocked(state == .loading || state == .failed)
    }
}
